### PR TITLE
feat: app.beforeStart support async function same as beforeClose

### PR DESCRIPTION
**Pull request by popomore**
##### Checklist

- [x] `npm test` passes
- [x] tests and/or benchmarks are included
- [x] commit message follows commit guidelines

##### Affected core subsystem(s)

app.beforeStart 

##### Description of change

app.beforeStart support async function same as beforeClose

### CODE REVIEW

File: lib/egg.js [hunk 1]
```diff
@@ -139,11 +139,11 @@ class EggCore extends KoaApplication {
   /**
    * excute scope after loaded and before app start
    *
-   * @param  {GeneratorFunction} scope function will excute before app start
+   * @param  {Function|GeneratorFunction|AsyncFunction} scope function will excute before app start
    */
   beforeStart(scope) {
```
Comment (shaoshuai0102) [inline lib/egg.js]: 对了 这个变量为啥命名为 scope，有点奇怪

File: lib/egg.js [hunk 2]
```diff
@@ -152,7 +152,9 @@ class EggCore extends KoaApplication {
 
     // ensure scope excutes after load completed
     process.nextTick(() => {
-      co(scope).then(() => done(), done);
+      co(function* () {
+        yield utils.callFn(scope);
+      }).then(() => done(), done);
```
Comment (shaoshuai0102) [inline lib/egg.js]: 上面没有 return，resolve 的第一个参数永远是 undefined，不需要 `() => done` 吧，直接 `then(done, done)` 就好

**Comment by popomore** [comment]
我一起发个版本

**Comment by popomore** [comment]
+ egg-core@1.8.0

File: lib/egg.js (hunk 2 above)
Comment (popomore) [inline lib/egg.js]: 哦，也是

**Pull request merged**